### DISCUSSION TITLE
Hydrogen loss in ring closure reaction when applying RunReactants

### DISCUSSION THREAD

**Discussion by xujiaboalex** [rdkit/rdkit#6095] [FAQ]
I write a reaction template SMARTS for indole synthesis, and the code is as follows:

```
def show_reaction(smarts, smiles: str):
    smiles, product_gt_smi = smiles.split('>>')
    smiles = smiles.split('.')

    mols_raw = [Chem.MolFromSmiles(smi) for smi in smiles]
    smiles_with_H = [Chem.MolToSmiles(mol, allHsExplicit=True) for mol in mols_raw]
    mols_with_H = [Chem.MolFromSmiles(smi) for smi in smiles_with_H]

    reaction = AllChem.ReactionFromSmarts(smarts)
    results = reaction.RunReactants(mols_with_H)
    product_pred = results[0][0]
    product_gt = Chem.MolFromSmiles(product_gt_smi)
    
    img = show_mol_grid(mols_with_H + [product_gt, product_pred], texts=['r1', 'r2', 'p_gt', 'p_pred'])

    display(img)
    product_gt_smi = Chem.MolToSmiles(product_gt,canonical=True)
    product_pred_smi = Chem.MolToSmiles(product_pred,canonical=True)
    print('product_gt:',product_gt_smi)
    print('product_pred:',product_pred_smi)


# my test case
smiles = 'IC1=CC=CC=C1N.[H]C#CC>>CC2=CC3=CC=CC=C3N2'
smarts = '[c:1]([#35,#53;D1;$(*c1ccccc1)])[c:2](-[#7A:3]).[#6;A;D1:5]#[#6;A;$(C-[#6,#14]):4]>>[c:1]1[c:5][c:4][n:3][c:2]1'

show_reaction(smarts, smiles)
```
However, when running test case, the hydrogen attached to indole nitrogen is missing. In addition, the SMILES of predicted product is `Cc1cc2ccccc2n1`, which the correct answer is `Cc1cc2ccccc2[nH]1`

Does anyone know how to solve it?

<img width="905" alt="截屏2023-02-17 17 46 38" src="https://user-images.githubusercontent.com/41980769/219610381-d2556054-0415-49ac-9eca-c5216d267052.png">

**Comment by bp-kelley** [comment]
Hydrogens are mostly implicit in smiles except when they are not.  In this case, you need to explicitly declare the hydrogen on the nitrogen on the product: [c:1]1[c:5][c:4][nH:3][c:2]1

```
smarts = '[c:1]([#35,#53;D1;$(*c1ccccc1)])[c:2](-[#7A:3]).[#6;A;D1:5]#[#6;A;$(C-[#6,#14]):4]>>[c:1]1[c:5][c:4][nH:3][c:2]1'
```

**Reply by xujiaboalex** [reply]
the reason I do not write explict hydrogen on the nitrogen is that sometimes a R group may attach to the nitrogen, for example a methyl group. Is there any way to combine these to cases, I mean, nitrogen has one nitrogen or a R group?
